### DISCUSSION TITLE
使用before_Link自定义链接顺序并未奏效

### DISCUSSION THREAD

**Discussion by Jackey990709** [xmake-io/xmake#5181] [Q&A] [answered]
我希望通过使用before_Link对链接顺序进行自定义，目的是在链接的过程中先链接Exception.c.o再链接startup_S32K146.S.o，但使用之后并没有实现预期的效果，依然先链接的startup_S32K146.S.o，不知是否是我的before_Link里编写的内容有问题
![飞鸽截图20240603105802](https://github.com/xmake-io/xmake/assets/147250832/2d75787d-86dc-4c1b-9bdd-babd141063c9)

**Comment by waruqi** [comment]
这不应该 链接脚本改该干的事情么。为啥不直接用  link_scripts

**Reply by Jackey990709** [reply]
能详细说说用 link_scripts的实现方法吗，在群里咨询，建议用的是on_link，但具体用法不知如何去写

**Reply by waruqi** [reply]
自己 google 下，那是 gcc 的东西，https://home.cs.colorado.edu/~main/cs1300/doc/gnu/ld_3.html

**Reply by Jackey990709** [reply]
请问我写了一个这样的linker.script，该如何应用在xmake中呢？
![飞鸽截图20240603172505](https://github.com/xmake-io/xmake/assets/147250832/23e02cff-d8b5-400c-a15e-511be111b587)

**Reply by star-hengxing** [reply]
直接 add files 或者 add ldflags 传入

**Reply by waruqi** [reply]
`add_files("xxx.lds")`

**Reply by Jackey990709** [reply]
我使用这个链接脚本并没有成功，是否有其他方式，在xmake.lua中能实现自定义link order吗

**Reply by waruqi** [reply]
那就你之前的方式 before_link 自己插入。。之前写法不对，应该直接改 objectfiles 不要走 target:set() 。。只有 scope api 才有对应 target:set

**Reply by Jackey990709** [reply]
我写了一个如图的before_link,执行xmake后报错error: method 'objectfiles' is not callable (a table value)
![飞鸽截图20240604141658](https://github.com/xmake-io/xmake/assets/147250832/bf18aaa3-d408-48f5-af75-073d70a1e0c4)

**Reply by Jackey990709** [reply]
可以帮忙看一下我这个脚本编写的哪里有问题导致报错吗，麻烦了

**Reply by waruqi** [reply]
哎，多翻翻源码，好多例子了。。https://github.com/xmake-io/xmake/blob/0f43f1c1503410ad099420954e6f4a701e885f60/xmake/rules/lex_yacc/yacc/xmake.lua#L45

target.objectfiles 是 function 你改写这个干嘛。。

**Reply by star-hengxing** [reply]
`target:objectfiles() = xxx`

**Reply by waruqi** [reply]
> `target:objectfiles() = xxx`

没这种语法

**Comment by star-hengxing** [comment]
看了下源码，objectfiles 接口返回的是 target 里的变量（如果有缓存），这里可以像指针一样直接修改内容就生效？

**Reply by waruqi** [reply]
缓存后，就一 table ，直接改就行了

**Comment by star-hengxing** [comment, marked as answer]
@Jackey990709 

```lua
target("main")
    set_kind("binary")
    add_files("src/*.c")

    before_link(function (target)
        local objectfiles = target:objectfiles()
        local tmp = objectfiles[1]
        objectfiles[1] = objectfiles[2]
        objectfiles[2] = tmp
    end)
```

这里当成指针用应该就行了
